### ISSUE TITLE
[Client Routing] Auto-reload page upon new deployed static assets

### ISSUE THREAD

**Issue by pieperz** [brillout/vite-plugin-ssr#292]
From Discord: Hey everyone, I've been seeing some "weird" script loading errors in Production and finally figured out the issue and wanted to see if anyone had a solution.  The issue starts each time we push to Production all the assets get a hash post-fix for example "vendor.abc123.js" then on the next push to production the post-fix will change to "vendor.zyx987.js". The problem is if someone has a tab open for the deployment of "vendor.abc123.js" and leaves the tab open. Then we deploy again and assets change to "vendor.zyx0987.js". Now when the user click or try's to use the tab with "vendor.abc123.js" they are met with a 404 can't find asset, which makes since because the new asset url "vendor.zyx987.js". This would be less of an issue if we were not using client side navigation because a full page refresh will refresh the asset url.

@brillout's response  (FYI the idea is generate a static file dist/client/assets/_buildId.js which the client regularly pings to check whether it is out-of-sync.)

**Comment by brillout** [comment]
A basic idea is to opt-out of Client Routing as soon as an asset could not be loaded. The current page transition (that would normally fail) falls back to Server Routing leading to a full reload and the latest version of the website is loaded.

If it works, it's then an easy strategy that, AFAICT, works for most use cases.

I'll have to experiment. (There are a couple of open questions to answer.)

**Comment by pieperz** [comment]
That could work for most use cases, however I think we would still have issues if the user clicks a button such as "Add To Cart". Often times this is the next action on a page that has been open for a long time.

@patryk-smc was looking around and found this issue in next.js https://github.com/vercel/next.js/issues/5652

```
// this goes in /pages/_app.tsx
const [buildId] = useBuildId(); // useSWR under the hood
const prevBuildId = usePrevious(buildId);
React.useEffect(() => {
    if (prevBuildId && buildId && prevBuildId !== buildId) {
        Router.reload();
    }
}, [buildId, prevBuildId]);
```

And I have an endpoint /api/build-id
```
const buildId = (req, res) => {
    res.status(200).json({ buildId: process.env.VERCEL_GIT_COMMIT_SHA });
};
export default buildId;
```

It would be nice to add in a popup telling them more info about why the page is about to refresh.

**Comment by patryk-smc** [comment]
I prefer an "invisible" approach, where we are just performing a full page reload, as Rom says. Majority of internet users are not used to see "update popup" on the website. I find that weird, actually.

**Comment by brillout** [comment]
I also think it should be invisible, whenever possible.

The best solution I can think so far of is to have a WebSocket (or polling for SSG) that listens to the build ID and then disables Client Routing. But this may induce increased load on the server, depending on how many users you expect.

> issues if the user clicks a button such as "Add To Cart".

That's usually not a problem since it doesn't induce a new asset to be loaded.

However, the app will break if you dynamically load components.

For dynamically loaded components I don't see an alternative than to do a full reload interrupting the user in whatever he is doing. In such situation a pop-up could make sense, since we cannot gracefully recover.

Another edge case: the server is re-deployed while a user is loading the website. A fallback to do a full reload would be nice.

I think we should ignore some of these edge cases.

**Comment by brillout** [comment]
One thing I can quickly implement (after the 0.4 release) is to add a `/dist/client/_build-id`. Not sure how much it would help you though. You could implement this pop-up, but it's not the best DX.

I'm thinking, I can also expose a new option to dynamically disable Client Routing. Should be easy.

So you can do the polling yourself and then disable Client Routing yourself.

I expect this to be quick to implement on my side.

**Comment by pieperz** [comment]
Any idea what the timeline on this is? Now that I have real customers on the site I'm seeing errors around this much more frequently.

**Comment by brillout** [comment]
I need your input here. Read my answers.

> add a `/dist/client/_build-id`

I can release this on the `0.4` beta, maybe on the `0.3` branch as well, let's see.

It should, AFAICT, be quick to implement.

**Comment by patryk-smc** [comment]
I have been looking at Next.js and how they are handling this.

They do not do anything when there are no errors. So if I have an old deployment open and I navigate to the page that got text updated in the new deployment - I will still see old text.

But if that page has a link in it, and router fails to prefetch it (due to 404), clicking that link will load that URL without client routing.

Currently, in that scenario, vite-plugin-ssr throws an error in the console. Could you @brillout make it load the URL without client routing too?

I'm not sure if that would solve all the problems, though. But it seems like a good first step. Then we probably should be able to handle other edge case with use of build ID. What do you guys think?

**Comment by brillout** [comment]
@patryk-smc I agree, that's a good plan.

1. Load the new page with server routing if some asset fails to load.
2. Expose a build ID.

We first experiment with 1. and see how it goes. Then 2. if it's not enough.

The main question I've with 1. is how reliable it is. We may reload the wrong URL. It needs to be tested.

Note that the URL of the new page will most of the time not 404 but it's one of its assets that will 404. So it's not as easy as it seems.

I'll work on this after I'm done with a couple of high prio things I'm currenlty working on. ETA for starting to work on this being next week. Feel free to share a specific timeline, if you need one.

**Comment by brillout** [comment]
Implemented a solution which I will release tomorrow. Preliminary tests indicate that it works for JavaScript and CSS assets but for image assets there still could be some problems. Let's see.

You'll have to migrate to 0.4 but I'll finish the migration docs.

If this approach doesn't work out, we can use the Next.js approach which is easy to implement.

For future reference, there is no global event listener for resources failing to load https://stackoverflow.com/questions/19098450/is-there-a-global-approach-to-catch-network-errors-in-javascript.

**Comment by patryk-smc** [comment]
Awesome. We are using almost zero images, so it shouldn't be a problem.

**Comment by brillout** [comment]
Fix released in `0.4.0-beta.31`. I'll leave this ticket open for a little while and close it if nothing comes up.

**Comment by brillout** [comment]
I guess we can close this.
